numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -99,16 +99,42 @@ _umath_strings_richcompare(
         PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip);
 
 
-static PyObject *
-set_legacy_print_mode(PyObject *NPY_UNUSED(self), PyObject *args)
-{
-    if (!PyArg_ParseTuple(args, "i", &npy_thread_unsafe_state.legacy_print_mode)) {
-        return NULL;
+NPY_NO_EXPORT int
+get_legacy_print_mode(void) {
+    /*
+     * Get the C value of the legacy printing mode.
+     * accessible from C. For simplicity the mode is encoded as an
+     * integer where INT_MAX means no legacy mode, and '113'/'121'/'125'
+     * means 1.13/1.21/1.25 legacy mode; and 0 maps to INT_MAX. We can
+     * upgrade this if we have more complex requirements in the future.
+     */
+    PyObject *val = NULL;
+    PyContextVar_Get(npy_static_pydata._format_options, NULL, &val);
+    if (val == NULL) {
+        PyErr_SetString(PyExc_SystemError,
+                        "NumPy internal error: unable to get _format_options "
+                        "context variable");
+        return -1;
     }
-    if (!npy_thread_unsafe_state.legacy_print_mode) {
-        npy_thread_unsafe_state.legacy_print_mode = INT_MAX;
+    PyObject *legacy_print_mode = NULL;
+    if (PyDict_GetItemStringRef(val, "legacy", &legacy_print_mode) == -1) {
```

Comment:
The string creation likely has a surprisingly large overhead.  If you worry about micro-optimizing, it probably makes sense to put this into a global and use `Py_Dict_GetItemRef` instead.

I think that micro-optimization may make sense, but the tiny overhead of the context-var lookup should not be large compared to building the result string for scalars even.

Reply:
Good call, I'll intern the string.